snapshotter/cmd/ctr-remote/commands/rpull.go

```diff
@@ -0,0 +1,162 @@
+		if desc.MediaType != containerdimages.MediaTypeDockerSchema1Manifest {
+			fmt.Printf("fetching %v... %v\n", desc.Digest.String()[:15], desc.MediaType)
+		}
+		return nil, nil
+	})
+
+	log.G(pCtx).WithField("image", ref).Debug("fetching")
+	labels := commands.LabelArgs(config.Labels)
+	if _, err := client.Pull(pCtx, ref, []containerd.RemoteOpt{
+		containerd.WithPullLabels(labels),
+		containerd.WithResolver(config.Resolver),
+		containerd.WithImageHandler(h),
+		containerd.WithSchema1Conversion,
+		containerd.WithPullUnpack,
+		containerd.WithPullSnapshotter("cc-snapshotter"),
+		containerd.WithImageHandlerWrapper(appendInfoHandlerWrapper(ref)),
+	}...); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+const (
+	// targetRefLabel is a label which contains image reference and will be passed
+	// to snapshotters.
+	targetRefLabel = "containerd.io/snapshot/cri.image-ref"
+	// targetManifestDigestLabel is a label which contains manifest digest and will be passed
+	// to snapshotters.
+	targetManifestDigestLabel = "containerd.io/snapshot/cri.manifest-digest"
+	// targetLayerDigestLabel is a label which contains layer digest and will be passed
+	// to snapshotters.
+	targetLayerDigestLabel = "containerd.io/snapshot/cri.layer-digest"
+	// targetImageLayersLabel is a label which contains layer digests contained in
+	// the target image and will be passed to snapshotters for preparing layers in
+	// parallel. Skipping some layers is allowed and only affects performance.
+	targetImageLayersLabel = "containerd.io/snapshot/cri.image-layers"
+)
+
+// appendInfoHandlerWrapper makes a handler which appends some basic information
+// of images like digests for manifest and their child layers as annotations during unpack.
+// These annotations will be passed to snapshotters as labels. These labels will be
+// used mainly by stargz-based snapshotters for querying image contents from the
+// registry.
+func appendInfoHandlerWrapper(ref string) func(f containerdimages.Handler) containerdimages.Handler {
+	return func(f containerdimages.Handler) containerdimages.Handler {
+		return containerdimages.HandlerFunc(func(ctx context.Context, desc imagespec.Descriptor) ([]imagespec.Descriptor, error) {
+			children, err := f.Handle(ctx, desc)
+			if err != nil {
+				return nil, err
+			}
+			switch desc.MediaType {
+			case imagespec.MediaTypeImageManifest, containerdimages.MediaTypeDockerSchema2Manifest:
+				for i := range children {
+					c := &children[i]
+					if containerdimages.IsLayerType(c.MediaType) {
+						if c.Annotations == nil {
+							c.Annotations = make(map[string]string)
+						}
+						c.Annotations[targetRefLabel] = ref
+						c.Annotations[targetLayerDigestLabel] = c.Digest.String()
+						c.Annotations[targetImageLayersLabel] = getLayers(ctx, targetImageLayersLabel, children[i:], labels.Validate)
+						c.Annotations[targetManifestDigestLabel] = desc.Digest.String()
+					}
+				}
+			}
+			return children, nil
+		})
+	}
+}
+
```

Comment:
Do we support non-dm-verity layers as well? Do we support ``runc`` or non-kata-cc Kata shim?